강의(교육) 신청 API 완성 #120

hot666666 opened the pull request:
### PR 타입(하나 이상의 PR 타입을 선택해주세요)
- [x] 기능 추가
- [ ] 기능 삭제
- [ ] 버그 수정
- [ ] 의존성, 환경 변수, 빌드 관련 코드 업데이트

### 반영 브랜치
hot666666/120-feat-education-application -> DOROEDU/develop

### 변경 사항
EducationApplication(교육 신청)  API를 위한 entity, DTO, mapper, service, repository 계층 작성

엔티티는 다음과 같음
![스크린샷 2023-11-30 오후 2 55 18](https://github.com/DOROEDU/Doro-Server/assets/73470656/6a42a1f7-42af-4a5b-bb46-8fdb194babf2)


### 테스트 결과
postman으로 테스트 완료

Ju-yeop committed:
Merge pull request #119 from DOROEDU/develop

Fix: FCM 및 Security 오류사항 수정

hot666666 committed:
Feat: 강의(교육) 신청 엔티티 #120
EducationApplication entity 작성
ClassGroup entity 작성

hot666666 committed:
Feat: 강의(교육) 신청 DTO 및 modelMapper 작성 #120
EducationApplicationMapper 및 DTO 생성

hot666666 committed:
Feat: 강의(교육) 신청의 학급정보 DTO 및 modelMapper 작성 #120
ClassGroupMapper 및 DTO 생성

hot666666 committed:
Feat: 강의(교육) 신청 서비스 및 레포지터리 작성 #120
EducationApplicationService
EducationApplicationRepository
ClassGroupService
ClassGroupRepository

hot666666 committed:
Feat: 강의(교육) 신청 서비스 예외 추가 #120
EducationApplicationService, ClassGroupService에서 발생할 수 있는 예외 Code에 추가

hot666666 committed:
Test: 강의(교육) 신청관련 서비스 및 mapper 테스트코드 작성 #120
EducationApplicationServiceTest
ClassGroupServiceTest
SpringBootTest를 이용한 mapper들 테스트는 통과해서, Disabled함

hot666666 committed:
Feat: 강의(교육) 신청 추가 API 작성
EducationApplication API

hot666666 committed:
Feat: 강의(교육) 신청 엔드포인트 시큐리티에 추가 #120
시큐리티 필터무시 설정에 "/apply/education/**" 추가

weightsforfun commented on src/main/java/com/example/DoroServer/domain/educationApplication/dto/EducationApplicationMapper.java:
```diff
@@ -0,0 +1,36 @@
+
+@Component
+@RequiredArgsConstructor
+public class EducationApplicationMapper {
```
modelMapper를 사용해서 mapping을 하셨군요!

저도 mapping을 어떤식으로 간소화 할까를 많이 고민했었어요.. 😂

mapstruct와 modelmapper 둘 중 하나를 골라야 했습니다
각자 장단점이 있었고 저는 mapstruct가 더 낫다고 생각해서 대부분은 mapstruct를 사용했는데
업데이트에 한해서는 setter 없이는 업데이트가 불가능해서 modelmapper를 사용했습니다!

지금도 충분히 좋은 방식이지만 2 mapper의 차이가 궁금하시면 내용을 알아보시고 mapstruct도 사용해봐도 좋을거 같아요! 
사용하시다가 모르는 부분 있으면 comment 남겨주시면 됩니다!! 
고생했어요 자바가 천직이신거 같은데요?😎 👍👍

hot666666 replied:
어맛.. 찾아보기 귀찮아서 간단해보이는 modelMapper를 사용했는데 딱 걸려부렸네요..😂👍👍
한번 나중에 찾아볼게요.. 왜냐하면 당신이 하는말에 전적으로 I'm 신뢰에요😎~

weightsforfun commented on src/main/java/com/example/DoroServer/domain/educationApplication/api/EducationApplicationApi.java:
```diff
@@ -0,0 +1,98 @@
+        return SuccessResponse.successResponse(updatedApplication);
+    }
+
+    @ApiOperation(value = "신청한 교육 삭제", notes = "신청한 교육을 삭제합니다. 학급정보도 같이 삭제됩니다.")
+    @DeleteMapping("/{applicationId}")
+    public SuccessResponse<?> delete(@PathVariable Long applicationId) {
+        applicationService.delete(applicationId);
+        return SuccessResponse.successResponse("신청이 삭제되었습니다.");
+    }
```
저번에 했던 얘기랑 같은 결의 얘기인데 만약 다른 사용자가 아무 id나 delete 요청을 하면 인가 로직이 없어서 작동될거 같아요! 

인가 로직이 있으면 확실히 편하긴 할텐데 이걸 어떻게 풀지 같이 고민해봐야 할거 같네요...

지금 당장 생각나는 것은
- 핸드폰 번호도 같이 받아서 일치 여부 확인
- 핸드폰으로 인증 번호 보내서 받기 (인증번호는 내부에서 만들어서 redis에 저장)

정도네요! 생각해 보시고 comment 주세요..!

hot666666 replied:
저는 핸드폰 인증을 하면, 토큰을 얻어 처리하는 방법으로 생각하고 있어요!

토큰으로 핸드폰 번호를 담아 서버의 비밀키로 서명을 하고 일정 시간동안 사용 가능한 JWT를 이용한다면

서버에선 해당 토큰을 통해 얻은 핸드폰 번호를 참고하여 동작하면 될 것 같습니다

근데 그렇게되면 JWT decode를 통해 누구나 핸드폰 번호를 볼 수 있기에 조금 찝찝할 수 있습니다(토큰이 유출되지 않는다면 문제 없지 않을까요?)

weightsforfun replied:
그것도 좋은방법 같아요! 코드를 좀 더 추가해야 하지만 가능하시다면 그렇게 하면 될거 같습니당 👍

hot666666 replied:
기존 휴대폰 인증 관련 코드를 봤는데, 그냥 조금 더 간단하게 세션방식으로 처리하는게 편할 것 같아요

- 핸드폰 인증
- 유효 5분인 key:세션ID, value:핸드폰번호 으로 redis에 저장
- 해당 세션 아이디를 핸드폰 인증한 유저한테 제공

weightsforfun replied:
넵! issue 등록 봤습니다! 브랜치 따서 개발하시고 mr 주세요 👍

weightsforfun commented on src/main/java/com/example/DoroServer/domain/educationApplication/entity/EducationApplication.java:
```diff
@@ -0,0 +1,53 @@
+    // == 연관관계 매핑 ==//
+
+    // EducationApplication과 ClassGroup은 일대다(One-to-Many) 관계
+    @Builder.Default
```
혹시 이 어노테이션은 왜 붙이신 건지 알 수 있을까요??

모르는 어노테이션이라 물어봅니다... 헤헤.. 😎

hot666666 replied:
빌더를 사용할 때, 해당필드에 대해서 기본 값을 설정하기 위해 사용했습니다🥹

weightsforfun replied:
오홍.... 좋은 기능 알아갑니다 👍👍👍

weightsforfun commented on src/main/java/com/example/DoroServer/domain/educationApplicationClassGroup/entity/ClassGroup.java:
```diff
@@ -0,0 +1,53 @@
+    private String className; // 학급 이름
+    private String educationConcept; // 교육 컨셉
+    private int numberOfStudents; // 학생 수
+    @ElementCollection
+    private List<LocalDate> educationDates; // 교육 날짜
```
저도 이부분을 많이 고민했었는데 만약 list가 길지 않으면 converter를 만들어서 string으로 저장하면 어떨까? 생각도 했어요 RDB를 사용하다보면 어떤 언어든 마주하는 상황이니까 hot66666님도 고민해 보시고 의견 남겨주시면 감사하겠습니다..! 👍

weightsforfun commented on src/test/java/com/example/DoroServer/domain/educationApplicationClassGroup/dto/ClassGroupMapperTest.java:
```diff
@@ -0,0 +1,170 @@
+    private ClassGroupReq setUpClassGroupReq() {
+        return ClassGroupReq.builder()
+                .className("1학년 1반")
+                .educationConcept("Q* 알고리즘")
+                .numberOfStudents(15)
+                .educationDates(List.of(
```
Q* 알고리즘은 1학년이 배우기 좀 어려운 알고리즘 같아요..!
테스트 케이스가 불편해요..! 👌

hot666666 replied:
냉장고등학교에 한계란 없습니다

weightsforfun commented on src/test/java/com/example/DoroServer/domain/educationApplicationClassGroup/service/ClassGroupServiceTest.java:
```diff
@@ -0,0 +1,161 @@
+import com.example.DoroServer.domain.educationApplicationClassGroup.entity.ClassGroup;
+import com.example.DoroServer.domain.educationApplicationClassGroup.repository.ClassGroupRepository;
+
+@ExtendWith(MockitoExtension.class)
+public class ClassGroupServiceTest {
```
꼭 할 필요는 없지만 만약 해보고 싶다면 controller 단 테스트도 알아보시고 하면 나쁘지 않을거 같아요..!

물론 저도 안했습니다 ㅎㅎ.. 그래서 지금 고통 받고 있어요 😎

weightsforfun reviewed (commented):
고생하셨습니다! 형식대로 잘 짜주셨어요! 고칠것보단 이건 어떤지에 대한 의견을 많이 남겼습니다 ㅎㅎ
생각해보시고 commet 남겨주시고 baseEntity 추가나 클래스명 수정하고 말해주세요!!

hot666666 committed:
Fix: 강의(교육) 신청 관련 엔티티 BaseEntity 상속 #120
EducationApplication, ClassGroup에서 BaseEntity 상속하도록 수정

hot666666 committed:
Fix: 강의(교육) 신청 엔드포인트 수정 #120
"apply/education" -> "/education-application"

hot666666 committed:
Refact: 강의(교육) 신청 테스트 리팩토링 #120
기존 테스트에 사용되던 setup을 모아 리팩토링
- EducationApplicationTestSetup
- ClassGroupTestSetup

hot666666 committed:
Chore: jackson-datatype-jsr310 의존성 추가 #120
EducationApplicationApi 슬라이스 테스트 진행 시, DateTime 타입을 처리하는데 사용

hot666666 committed:
Test: 강의(교육) 신청 Api 테스트 코드 작성 #120
EducationApplication Api 테스트 코드 작성

weightsforfun commented:
일단 이정도로 merge 하고 고도화 작업으로 새로운 이슈 여실래요? 민혁쌤이랑 얘기 잘 하고 있나요??

hot666666 commented:
고도화 작업이면 위에서 언급한 인증이랑 S3 이용한 이미지 업로드인가요?
